clippy_lints/src/doc.rs

```diff
@@ -609,3 +694,43 @@ fn check_word(cx: &LateContext<'_>, word: &str, span: Span) {
         );
     }
 }
+
+struct FindPanicUnwrap<'a, 'tcx> {
+    lcx: &'a LateContext<'tcx>,
+    typeck_results: &'tcx ty::TypeckResults<'tcx>,
+    found_panicking: bool,
+}
+
+impl<'a, 'tcx> Visitor<'tcx> for FindPanicUnwrap<'a, 'tcx> {
+    type Map = Map<'tcx>;
+
+    fn visit_expr(&mut self, expr: &'tcx Expr<'_>) {
+        // check for `begin_panic`
+        if_chain! {
+            if let ExprKind::Call(ref func_expr, _) = expr.kind;
+            if let ExprKind::Path(QPath::Resolved(_, ref path)) = func_expr.kind;
+            if let Some(path_def_id) = path.res.opt_def_id();
+            if match_panic_def_id(self.lcx, path_def_id);
+            then {
+                self.found_panicking = true;
+            }
+        }
+
+        // check for `unwrap`
+        if let Some(arglists) = method_chain_args(expr, &["unwrap"]) {
+            let reciever_ty = self.typeck_results.expr_ty(&arglists[0][0]).peel_refs();
+            if is_type_diagnostic_item(self.lcx, reciever_ty, sym::option_type)
+                || is_type_diagnostic_item(self.lcx, reciever_ty, sym::result_type)
+            {
+                self.found_panicking = true;
+            }
+        }
+
+        // and check sub-expressions
+        intravisit::walk_expr(self, expr);
+    }
+
+    fn nested_visit_map(&mut self) -> NestedVisitorMap<Self::Map> {
+        NestedVisitorMap::None
```

Comment:
You want to at least have [`NestedVisitorMap::OnlyBodies`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/intravisit/enum.NestedVisitorMap.html#variant.OnlyBodies) here, so that closures inside functions get visited.

Otherwise 

```rust
fn foo(opt: Option<u32>) {
    opt.map(|x| if x == 10 { panic!() });
}
```
won't get linted.

Please also add a test for this.

---

The question is, if you even want to use `NestedVisitorMap::All`, which would also lint inner functions. One argument in favor would be, that inner functions are called somewhere in the body (Otherwise rustc would emit an `unused` warning). The argument against this is to be conservative and have this as a potential FN.

Reply:
(I just copied and pasted this code, with a few adjustments, from one of the lints that found `todo!()` and things like that.) I'll do which ever one you thing is best.

Reply:
In that case, just use `OnlyBodies`, but also add a test for a nested function, that might panic.

Reply:
I suppose I should do the following?
> If for some reason you want the nested behavior, but don't have a Map at your disposal: then you should override the visit_nested_XXX methods, and override this method to panic!(). This way, if a new visit_nested_XXX variant is added in the future, we will see the panic in your code and fix it appropriately.
https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/intravisit/trait.Visitor.html#tymethod.nested_visit_map

Reply:
No you have a HIR map, but it is hidden in the `TyCtxt` of the `LateContext`.

```suggestion
        NestedVisitorMap::OnlyBodies(self.cx.tcx.hir())
```